optaplanner-core/src/main/java/org/optaplanner/core/api/score/constraint/ConstraintMatch.java

```diff
@@ -82,7 +83,12 @@ public int compareTo(ConstraintMatch other) {
             return constraintName.compareTo(other.constraintName);
         } else {
             for (int i = 0; i < justificationList.size() && i < other.justificationList.size(); i++) {
-                int comparison = ((Comparable) justificationList.get(i)).compareTo(other.justificationList.get(i));
+                Object left = justificationList.get(i);
+                Object right = other.justificationList.get(i);
+                Comparator comparator = left instanceof ConstraintJustification ?
```

Comment:
This confuses me. "left" is a problem fact or a planning entity.
This code reads: "if a problem fact is an instance of ConstraintJustification, it has a method that returns a comperator?

Reply:
Precisely. By implementing `ConstraintJustification`, you tell this code how to compare the instances without having the instances implement `Comparable<>`.